go/vt/vtgate/scatter_conn.go

```diff
@@ -366,9 +366,15 @@ func (stc *ScatterConn) StreamExecuteMulti(
 
 // Commit commits the current transaction. There are no retries on this operation.
 func (stc *ScatterConn) Commit(context context.Context, session *SafeSession) (err error) {
+	if session == nil {
+		return fmt.Errorf("cannot commit: empty session")
+	}
 	if !session.InTransaction() {
 		return fmt.Errorf("cannot commit: not in transaction")
 	}
+	if len(session.ShardSessions) == 0 {
+		return fmt.Errorf("cannot commit: invalid session, always update session from RPC response")
```

Comment:
Without the empty shardsessions check here, we wouldn't catch the behavior that motivated this fix - which is silent success on commit even though the client didn't update its session after a DML.